src/hilok.cpp

```diff
@@ -1,84 +1,130 @@
+        ++it;
+        if (m_shared || it != m_refs.end()) {
+            kref.m_mut->unlock_shared();
+        } else {
+            kref.m_mut->unlock();
+        }
+
+        m_mgr.erase_safe(kref);
+    }
+}
+
 HiHandle HiLok::read(std::string_view path, bool block, int timeout) {
     void *cur = nullptr;
     std::pair<void *, std::string> key;
-    std::vector<std::shared_timed_mutex *>refs;
+    std::vector<HiKeyRef> refs;
     try {
         for (auto it = PathSplit(path, m_sep); it != it.end(); ++it) {
             key = {cur, *it};
-            std::shared_timed_mutex *mut;
-            {
-                std::lock_guard<std::mutex> guard(m_mutex);
-                mut = &m_map[key];
-            }
+            std::shared_ptr<HiMutex> mut = _get_mutex(key);
             bool ok = shared_lock_with_params(*mut, block, timeout);
-            if (!ok)
+            if (!ok) {
+                mut.reset();
                 throw HiErr("failed to lock");
-            refs.push_back(mut);
-            cur = (void *) mut;
+            }
+            refs.emplace_back(mut, key);
+            cur = (void *) mut.get();
         }
     } catch (...) {
-        HiHandle(true, refs).release();
+        auto hh = HiHandle(*this, true, refs);
+        refs.clear();
+        hh.release();
         throw;
     }
-    return HiHandle(true, refs);
+    return HiHandle(*this, true, refs);
+}
+
+std::shared_ptr<HiMutex> HiLok::_get_mutex(std::pair<void *, std::string> key) {
+    std::lock_guard<std::mutex> guard(m_mutex);
+    auto it = m_map.find(key);
+    if (it == m_map.end()) {
+        return m_map[key] = std::make_shared<HiMutex>(m_recursive);
+    } else {
+        return it->second;
+    }
 }
 
+
 HiHandle HiLok::write(std::string_view path, bool block, int timeout) {
     void *cur = nullptr;
     std::pair<void *, std::string> key;
-    std::vector<std::shared_timed_mutex *>refs;
+    std::vector<HiKeyRef>refs;
     try {
         for (auto it = PathSplit(path, m_sep); it != it.end(); ) {
             key = {cur, *it};
+            std::shared_ptr<HiMutex> mut = _get_mutex(key);
+            
             ++it;
-            std::shared_timed_mutex *mut;
-            {
-                std::lock_guard<std::mutex> guard(m_mutex);
-                mut = &m_map[key];
-            }
             bool ok;
-
             if (it != it.end())
                 ok = shared_lock_with_params(*mut, block, timeout);
             else
                 ok = lock_with_params(*mut, block, timeout);
 
-            if (!ok)
+            if (!ok) {
+                mut.reset();
                 throw HiErr("failed to lock");
-            
-            refs.push_back(mut);
-            cur = (void *) mut;
+            }
+
+            refs.emplace_back(mut, key);
+            cur = (void *) mut.get();
         }
     } catch (...) {
-        HiHandle(true, refs).release();
+        auto hh = HiHandle(*this, true, refs);
+        refs.clear();
+        hh.release();
```

Comment:
that would be "guard" like behavior.   which would be a different interface.   i handle the "with block" stuff at the python level, not the C++ level.   i suppose we could change the interface to be "with block" at C++ level too, but i wasn't sure how best to deal with that since __enter__ and __exit__ are different endpoints.   i think we need to leave it.

Reply:
i think it can.

Reply:
but its super tricky.   the write() and read() functions would then have to be passed in the shared-pointer-to-themselves as an argument!

Reply:
still didnt work, because python's with block doesn't hold the ref